lib/perl/Genome/Qc/Result.pm

```diff
@@ -0,0 +1,146 @@
+package Genome::Qc::Result;
+
+use strict;
+use warnings;
+use Genome;
+
+class Genome::Qc::Result {
+    is => 'Genome::SoftwareResult::StageableSimple',
+    has_input => [
+        alignment_result => {
+            is => 'Genome::InstrumentData::AlignedBamResult',
+        },
+        config_name => {
+            is => 'Text',
+        },
+    ],
+};
+
+sub qc_config {
+    my $self = shift;
+    return Genome::Qc::Config->get(name => $self->config_name);
+}
+
+sub get_metrics {
+    my $self = shift;
+    my @metrics = Genome::SoftwareResult::Metric->get(
+        software_result => $self,
+    );
+    return map {$_->metric_name => $_->metric_value} @metrics;
+}
+
+sub _run {
+    my $self = shift;
+    my %tools = $self->_tools;
+    my $process_graph = Genome::WorkflowBuilder::StreamGraph->create(
+        output_xml => File::Spec->join($self->temp_staging_directory, "out.xml"),
+    );
+    my %process_ref;
+    while (my ($name, $tool) = each %tools) {
+        $process_ref{$name} = Genome::WorkflowBuilder::StreamProcess->create(
+            name => $name,
+            args => [$tool->cmd_line],
+            in_file_link => $self->_input_file_for_tool($tool, $name),
+            out_file_link => $self->_output_file_for_tool($name),
+            err_file_link => $self->_error_file_for_tool($name),
+        );
+        $process_graph->add_process($process_ref{$name});
+    }
+
+    while (my ($name, $tool) = each %tools) {
+        my $dependency = $self->_dependency_for_tool($name);
+        if (defined $dependency) {
+            my $link = Genome::WorkflowBuilder::StreamLink->create(
+                source => $process_ref{$dependency->{name}},
+                target => $process_ref{$name},
+                source_fd => $dependency->{fd},
+            );
+            $process_graph->add_link($link);
+        }
+    }
+    $process_graph->execute;
+
+    for my $tool (values %tools) {
+        $self->_add_metrics($tool);
+    }
+    return 1;
+}
+
+sub _dependency_for_tool {
```

Comment:
I'm not sure if its useful or not, but the lines:

```` perl
my ($self, $name) = @_;
my $something = $self->qc_config->get_commands_for_alignment_result->{$name}->some_key
if (defined $something) {
   return //some processing of the value
}
return undef;
````
Seem to appear in 4 or 5 places in this class. Maybe its worth extracting into some sort of helper?  I'm not sure its terribly useful at this point though as its only a small amount of duplication, and each case is slightly different. I had thought maybe something like:

````perl
sub _value_for_tool {
   my ($self, $name, $value, $callback) = @_;
   my $x = $self->qc_config->get_commands_for_alignment_result->{$name}->{$value};
   if (defined $x) {
      return $callback->($x);
   }
   return undef;
}

sub _error_file_for_tool {
   my ($self, $name) = @_;
   return $self->_value_for_tool(
    $name,
    'in_file',
    sub {  return File::Spec->join($self->temp_staging_directory, shift); }
   );
}
````
But after looking at it, I'm not sure that the loss of clarity is worth the deduplication. Your call!